perf/perf_stress_ng.py

```diff
@@ -0,0 +1,252 @@
+import time
+from avocado.utils import process
+from avocado import Test
+from avocado.utils import process, build, archive
+from avocado.utils.software_manager.manager import SoftwareManager
+
+
+def clear_dmesg():
+    process.run("dmesg -C ", sudo=True)
+
+
+def collect_dmesg(object):
+    return process.system_output("dmesg").decode()
+
+
+class Stressng(Test):
+
+    """
+    Perf record with CPU stress
+    :source: git://kernel.ubuntu.com/cking/stress-ng.git
+    :param cpu_per: CPU Percentage to load
+    """
+
+    def setUp(self):
+        smm = SoftwareManager()
+        self.timeout = self.params.get('timeout', default='1')
+        self.cpu_per = self.params.get("cpu_load", default='10')
+        self.profile_dur = int(
+            self.params.get(
+                "profile_duration",
+                default='1'))
+        run_type = self.params.get('type', default='upstream')
+        if run_type == "upstream":
+            asset_url = 'https://github.com/ColinIanKing/stress-ng/archive/master.zip'
+            tarball = self.fetch_asset(
+                'stressng.zip', locations=[asset_url], expire='7d')
+            archive.extract(tarball, self.workdir)
+            sourcedir = os.path.join(self.workdir, 'stress-ng-master')
+            os.chdir(sourcedir)
+            result = build.run_make(
+                sourcedir, process_kwargs={
+                    'ignore_status': True})
+            for line in str(result).splitlines():
+                if 'error:' in line:
+                    self.cancel(
+                        "Build Failed, Please check the build logs for details !!")
+            build.make(sourcedir, extra_args='install')
+            clear_dmesg()
+        elif run_type == "distro":
+            deps = ['gcc', 'perf', 'stress-ng']
+            for package in deps:
+                if not smm.check_installed(
+                        package) and not smm.install(package):
+                    self.cancel(
+                        '%s is needed for the test to be run' %
+                        package)
+
+    def is_process_running(self, process_name):
+        for proc in psutil.process_iter(['pid', 'name']):
+            if process_name in proc.info['name']:
+                return True
+        return False
+
+    def kill_process_by_name(self, process_name):
+        for proc in psutil.process_iter(['pid', 'name']):
+            if process_name in proc.info['name']:
+                try:
+                    process = psutil.Process(proc.info['pid'])
+                    process.terminate()
+                    self.log.info(f"Process  %s terminated" % (process_name))
+                except psutil.NoSuchProcess:
+                    self.log.info(f"Process  %s not found" % (process_name))
+                    return
+
+    def vmstat_result(self, filename):
+        var = 0
+        sum = 0
+        result = 0
+        cmd = "awk '!/procs|swpd/ && NF > 0 { print $15 }' %s > /tmp/data1.txt" % filename
```

Comment:
i know it is tricky but still can this will handle via python

Reply:
The code flow copies vmstat data to a file while perf data is captured and then tries to read vmstat data, add up all the wait time for comparison.    Since it is an comparision both vmstat and perf needs to be captured at the same time.

the limitation with Python script in this approach is,  it doesn't allow to convert string to int.   it complains of "invalid literal for int() with base 10: 'id'".    

 whereas awk is a more efficient way to handle these situations. Let me know if you agree/ have suggestions 
 
      input_file = open(filename, 'r')
     output_file = open("/tmp/data1.txt", 'w')
     for line in input_file:
         if 'procs' not in line or 'swpd' not in line:
             fields = line.strip().split()
             if len(fields) > 14:
                 val = int(fields[14])
                 output_file.write(val + '\n')
                 if var < 10:
                     var += 1
                     sum = sum + val
                 else:
                     sum = sum + val
                     result = 100 - sum // 11
                     var = 0
                     sum = 0

Reply:
Thanks @manvanthar , awk hold good for time been